### ISSUE TITLE
Different padding computation in ops.scale_masks versus letterbox transformation.

### ISSUE THREAD

**Issue by baptist** [ultralytics/ultralytics#20953]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Other

### Bug

Hi,
In the context of instance segmentation I noticed the following behavior:
The `LetterBox` transformation computes the padding (when it is needed to center) using the following computations:
(https://github.com/ultralytics/ultralytics/blob/main/ultralytics/data/augment.py#L1691)
```
top, bottom = int(round(dh - 0.1)) if self.center else 0, int(round(dh + 0.1))
left, right = int(round(dw - 0.1)) if self.center else 0, int(round(dw + 0.1))
```
However, when using the default SegmentationPredictor functionality to perform inference and the retina_masks equaling True (for more accurate masks) the `scale_masks` functionality computes padding as follows:
(https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/ops.py#L759)
```
top, left = (int(pad[1]), int(pad[0])) if padding else (0, 0)  # y, x
bottom, right = (int(mh - pad[1]), int(mw - pad[0]))
```
This in some cases leads to 1 pixel differences in the padding values leading to small shifts when upscaling masks to the original image sizes. In attachment an image with on the left side the masks using the original scale_masks function and on the right an adapted version (that uses the same padding computations as in the LetterBox and scale_boxes functions).

![Image](https://github.com/user-attachments/assets/a67fe089-431b-4ee3-8d59-15815402c726)

Best regards,
Baptist

### Environment

The code referred to above is in the master branch.

### Minimal Reproducible Example

See above.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @baptist, thank you for your detailed report and for highlighting this padding inconsistency in instance segmentation! 🧐 Your insights and attached example are very helpful for understanding the issue.

This is an automated response to let you know that an Ultralytics engineer will review and assist you soon.

If possible, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) to help us investigate further. Even with your current description and image, a concise code snippet or dataset sample that reliably reproduces the pixel shift would be valuable.

In the meantime, we recommend:
- Reviewing our [Docs](https://docs.ultralytics.com/) for relevant [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples.
- Upgrading to the latest `ultralytics` version and requirements in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in the following up-to-date, verified environments (all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Feel free to join the Ultralytics community for discussion and support:
- [Discord](https://discord.com/invite/ultralytics) 🎧
- [Discourse](https://community.ultralytics.com/)
- [Subreddit](https://reddit.com/r/Ultralytics)

Thank you for helping improve Ultralytics!

**Comment by Y-T-G** [comment]
Can you open a PR for this?

**Comment by Y-T-G** [comment]
@horsto Seems related to your issue

**Comment by horsto** [comment]
Yes, this very much looks like what I am seeing! @baptist - for reference #20918

**Comment by baptist** [comment]
> Can you open a PR for this?

I can't, it seems. Anyway the fix can be something like this:
```
def scale_masks(masks, shape, padding: bool = True):
    """
    Rescale segment masks to target shape.

    Args:
        masks (torch.Tensor): Masks with shape (N, C, H, W).
        shape (tuple): Target height and width as (height, width).
        padding (bool): Whether masks are based on YOLO-style augmented images with padding.

    Returns:
        (torch.Tensor): Rescaled masks.
    """
    if padding:
        mh, mw = masks.shape[2:]
        gain = min(mh / shape[0], mw / shape[1])  # gain  = old / new
        pad = [mw - shape[1] * gain, mh - shape[0] * gain]  # wh padding
        pad[0] /= 2
        pad[1] /= 2
        top, left = (int(round(pad[1] - 0.1)), int(round(pad[0] - 0.1)))
        bottom, right = (
            mh - int(round(pad[1] + 0.1)),
            mw - int(round(pad[0] + 0.1)),
        )
        masks = masks[..., top:bottom, left:right]

    masks = F.interpolate(masks, shape, mode="bilinear", align_corners=False)  # NCHW
    return masks
```
It seems scale_image needs to be adapted as well.

**Comment by horsto** [comment]
Thanks @baptist , I will give that a try asap. 
@Y-T-G Do you know which other function should be adapted in the same go? @baptist mentions `scale_image()` - anything else?

**Comment by Y-T-G** [comment]
Probably those two. `scale_masks` is used if you use `retina_mask`. Otherwise, `scale_image` is used.

**Comment by horsto** [comment]
I have opened a PR based on the code of @baptist here: https://github.com/ultralytics/ultralytics/pull/20957 
This seems to work!

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in [PR #20957](https://github.com/ultralytics/ultralytics/pull/20957): **ultralytics 8.3.152 Optimize padding for Segment mask processing**. The update improves mask scaling accuracy by refining how padding and cropping are handled in image processing functions. By updating `scale_image` and `scale_masks` to use `round()` for boundary calculations, the changes aim to deliver more precise mask alignment and reduce off-by-one errors—especially beneficial for segmentation edge cases.

Big thanks to @horsto for authoring this PR, with valuable contributions from @glenn-jocher, @Y-T-G, @Laughing-q, and @baptist.

To test if the fix resolves your issue, you can upgrade to the latest version with:
```
pip install -U ultralytics>=8.3.152
```
Alternatively, try the latest code directly from the main branch:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
You may also wait for the next official PyPI release if you prefer.

Please let us know if these changes resolve the issue or if any problems persist. Thanks again for reporting this, and any further feedback is always welcome! 🙏
